Update websocket/ws types for v3.0.0

elithrar opened the pull request:
Summary: Updates type definitions to reflect updated callback function signatures (e.g. adding `http.IncomingMessage` to the headers event) and additional client & server options.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/websockets/ws/blob/master/doc/ws.md
- [ ] Increase the version number in the header if appropriate. *Note*: no version number present.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

elithrar commented:
(Fixed issue in test file)

lpinca commented on types/ws/index.d.ts:
```diff
@@ -64,18 +66,22 @@ declare class WebSocket extends events.EventEmitter {
-    on(event: 'message', cb: (data: any, flags: { binary: boolean }) => void): this;
-    on(event: 'ping', cb: (data: any, flags: { binary: boolean }) => void): this;
-    on(event: 'pong', cb: (data: any, flags: { binary: boolean }) => void): this;
+    on(event: 'headers', cb: (headers: string[], request: http.IncomingMessage) => void): this;
```
The `headers` argument is an object: https://github.com/websockets/ws/blob/master/doc/ws.md#event-headers-1.

elithrar replied:
Fixed (original type def had this)

lpinca replied:
This is correct for the server not the client.

elithrar replied:
Noted, and corrected.

lpinca commented on types/ws/index.d.ts:
```diff
@@ -64,18 +66,22 @@ declare class WebSocket extends events.EventEmitter {
-    on(event: 'ping', cb: (data: any, flags: { binary: boolean }) => void): this;
-    on(event: 'pong', cb: (data: any, flags: { binary: boolean }) => void): this;
+    on(event: 'headers', cb: (headers: string[], request: http.IncomingMessage) => void): this;
+    on(event: 'message', cb: (data: any) => void): this;
```
Not sure how it can be specified but `data` can be one of `String`, `Buffer`, `ArrayBuffer`, `Buffer[]`.

elithrar replied:
`on(event: 'message', cb: (data: string|Buffer|ArrayBuffer|Buffer[]) => void): this;`

lpinca commented on types/ws/index.d.ts:
```diff
@@ -64,18 +66,22 @@ declare class WebSocket extends events.EventEmitter {
-    on(event: 'pong', cb: (data: any, flags: { binary: boolean }) => void): this;
+    on(event: 'headers', cb: (headers: string[], request: http.IncomingMessage) => void): this;
+    on(event: 'message', cb: (data: any) => void): this;
+    on(event: 'ping', cb: (data: any) => void): this;
```
`data` is a `Buffer`.

lpinca commented on types/ws/index.d.ts:
```diff
@@ -64,18 +66,22 @@ declare class WebSocket extends events.EventEmitter {
+    on(event: 'headers', cb: (headers: string[], request: http.IncomingMessage) => void): this;
+    on(event: 'message', cb: (data: any) => void): this;
+    on(event: 'ping', cb: (data: any) => void): this;
+    on(event: 'pong', cb: (data: any) => void): this;
```
`data` is a `Buffer`.

lpinca commented on types/ws/index.d.ts:
```diff
@@ -87,18 +93,22 @@ declare namespace WebSocket {
+        rejectUnauthorized?: boolean;
         passphrase?: string;
+        ciphers?: string;
         cert?: any;
```
`String|String[]|Buffer|Buffer[]` and same is  for `key` and `ca`.

elithrar replied:
Done

lpinca commented on types/ws/index.d.ts:
```diff
@@ -87,18 +93,22 @@ declare namespace WebSocket {
+        ciphers?: string;
         cert?: any;
+        key?: any;
+        pfx?: any;
```
`String|Buffer`.

elithrar replied:
Fixed

lpinca commented on types/ws/index.d.ts:
```diff
@@ -134,16 +144,19 @@ declare namespace WebSocket {
         close(cb?: (err?: any) => void): void;
         handleUpgrade(request: http.IncomingMessage, socket: net.Socket,
             upgradeHead: Buffer, callback: (client: WebSocket) => void): void;
+        shouldHandle(request: http.IncomingMessage): void
```
Return value is `boolean`.

elithrar replied:
Fixed

elithrar commented:
Addressed comments (updated incorrect types from original submission)

lpinca commented on types/ws/index.d.ts:
```diff
@@ -87,18 +93,22 @@ declare namespace WebSocket {
-        ca?: any[];
         ciphers?: string;
-        rejectUnauthorized?: boolean;
+        cert?: any;
```
This is also `string|string[]|Buffer|Buffer[]` also I _think_ string should in lowercase? In that case it should be fixed below.

elithrar replied:
Fixed.

lpinca commented on types/ws/index.d.ts:
```diff
@@ -134,16 +144,19 @@ declare namespace WebSocket {
-        addListener(event: 'headers', cb: (headers: string[]) => void): this;
         addListener(event: 'connection', cb: (client: WebSocket) => void): this;
+        addListener(event: 'error', cb: (err: Error) => void): this;
+        addListener(event: 'headers', cb: (headers: {}, request: http.IncomingMessage) => void): this;
```
`headers: string[]`

elithrar replied:
Fixed.

lpinca commented on types/ws/ws-tests.ts:
```diff
@@ -38,16 +41,16 @@ var WebSocketServer = WebSocket.Server;
 
-    wsc.on('message', (data, flags) => {
-        console.log('Roundtrip time: ' + (Date.now() - parseInt(data)) + 'ms', flags);
+    wsc.on('message', (data: any) => {
```
Is `any` correct here?

elithrar replied:
Correct: should be explicit here.

Package users can alternatively specify the union types—`wsc.on('message', (data: string|Buffer|ArrayBuffer|Buffer[]) => { ... }`—and then `typeof message` in a switch/if-else to handle each possible type as (if) needed.

lpinca commented:
@elithrar thank you for this.

lpinca commented:
On line 27 there is `upgradeReq: http.IncomingMessage;`. This property has been removed in `ws@3`.

elithrar commented:
@lpinca Noted.

* Fixed this (removed)
* Added extensions (right as you edited this)
 * Fixed `protocol` to be of type `number` (was `string` in the previous type decl)
* Sorted the properties in the `WebSocket` class to align with the docs (others left as-is, as I have manually sorted them to align with the docs)

lpinca commented on types/ws/index.d.ts:
```diff
@@ -11,21 +12,22 @@ import * as http from 'http';
-    protocolVersion: string;
-    url: string;
-    supports: any;
     upgradeReq: http.IncomingMessage;
```
Sorry to pester but this is no longer available in `ws@3`.

elithrar commented:
Made a couple of minor additions:

* Created the `WebSocket.Data` type - `type Data = string | Buffer | ArrayBuffer | Buffer[];` - that represents the union type for a message payload. This makes it easier (shorter!) for package users to match the type signature by writing `(message: WebSocket.Data)`.
* A `CertMeta` type - `type CertMeta = string | string[] | Buffer | Buffer[];` - a similar convenience when providing certificate-related options.

This also makes maintaining the type definitions easier in the future.

andy-ms commented:
Thanks!
